// ionic-audio/dist/ionic-audio.module.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IonicModule } from 'ionic-angular';
import { AudioTrackComponent } from './ionic-audio-track-component';
import { AudioTrackProgressComponent, AudioTrackProgressBarComponent } from './ionic-audio-track-progress-component';
import { AudioTrackPlayComponent } from './ionic-audio-track-play-component';
import { AudioTimePipe } from './ionic-audio-time-pipe';
import { AudioProvider, WebAudioProvider } from './ionic-audio-providers';
var IonicAudioModule = IonicAudioModule_1 = (function () {
    function IonicAudioModule() {
    }
    /**
     * Configures Ionic Audio to use either Cordova or HTML5 audio.
     * If no ```audioProvider``` is set it will automatically choose one based on the current environment
     */
    IonicAudioModule.forRoot = function (audioProvider) {
        return {
            ngModule: IonicAudioModule_1,
            providers: [
                audioProvider || { provide: AudioProvider, useClass: WebAudioProvider }
            ]
        };
    };
    return IonicAudioModule;
}());
IonicAudioModule = IonicAudioModule_1 = __decorate([
    NgModule({
        imports: [CommonModule, IonicModule],
        declarations: [
            AudioTrackComponent,
            AudioTrackProgressComponent,
            AudioTrackProgressBarComponent,
            AudioTrackPlayComponent,
            AudioTimePipe
        ],
        exports: [
            AudioTrackComponent,
            AudioTrackProgressComponent,
            AudioTrackProgressBarComponent,
            AudioTrackPlayComponent,
            AudioTimePipe
        ],
        providers: [],
        schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
], IonicAudioModule);
export { IonicAudioModule };
var IonicAudioModule_1;
