// ionic-audio/dist/ionic-audio-web-track.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Injectable, Optional } from '@angular/core';
window.AudioContext = window['AudioContext'] || window['webkitAudioContext'];
/**
 * Creates an HTML5 audio track
 *
 * @export
 * @class WebAudioTrack
 * @constructor
 * @implements {IAudioTrack}
 */
var WebAudioTrack = (function () {
    function WebAudioTrack(src, preload) {
        // audio context not needed for now
        // @Optional() private ctx: AudioContext = undefined
        // this.ctx = this.ctx || new AudioContext();
        if (preload === void 0) { preload = 'none'; }
        this.src = src;
        this.preload = preload;
        this.isPlaying = false;
        this.isFinished = false;
        this._progress = 0;
        this._completed = 0;
        this.createAudio();
    }
    WebAudioTrack.prototype.createAudio = function () {
        var _this = this;
        this.audio = new Audio();
        this.audio.src = this.src;
        this.audio.preload = this.preload;
        //this.audio.controls = true;
        //this.audio.autoplay = false;
        this.audio.addEventListener("timeupdate", function (e) { _this.onTimeUpdate(e); }, false);
        this.audio.addEventListener("error", function (err) {
            console.log("Audio error => track " + _this.src, err);
            _this.isPlaying = false;
        }, false);
        this.audio.addEventListener("canplay", function () {
            console.log("Loaded track " + _this.src);
            _this._isLoading = false;
            _this._hasLoaded = true;
        }, false);
        this.audio.addEventListener("playing", function () {
            console.log("Playing track " + _this.src);
            _this.isFinished = false;
            _this.isPlaying = true;
        }, false);
        this.audio.addEventListener("ended", function () {
            _this.isPlaying = false;
            _this.isFinished = true;
            console.log('Finished playback');
        }, false);
        this.audio.addEventListener("durationchange", function (e) {
            _this._duration = e.target.duration;
        }, false);
    };
    WebAudioTrack.prototype.onTimeUpdate = function (e) {
        if (this.isPlaying && this.audio.currentTime > 0) {
            this._progress = this.audio.currentTime;
            this._completed = this.audio.duration > 0 ? Math.trunc(this.audio.currentTime / this.audio.duration * 100) / 100 : 0;
        }
    };
    WebAudioTrack.formatTime = function (value) {
        var s = Math.trunc(value % 60);
        var m = Math.trunc((value / 60) % 60);
        var h = Math.trunc(((value / 60) / 60) % 60);
        return h > 0 ? (h < 10 ? '0' + h : h) + ":" + (m < 10 ? '0' + m : m) + ":" + (s < 10 ? '0' + s : s) : (m < 10 ? '0' + m : m) + ":" + (s < 10 ? '0' + s : s);
    };
    Object.defineProperty(WebAudioTrack.prototype, "id", {
        /**
         * Gets the track id
         *
         * @property id
         * @type {number}
         */
        get: function () {
            return this._id;
        },
        /**
         * Sets the track id
         *
         * @property id
         */
        set: function (v) {
            this._id = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebAudioTrack.prototype, "duration", {
        /**
         * Gets the track duration, or -1 if it cannot be determined
         *
         * @property duration
         * @readonly
         * @type {number}
         */
        get: function () {
            return this._duration;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebAudioTrack.prototype, "progress", {
        /**
         * Gets current track time (progress)
         *
         * @property progress
         * @readonly
         * @type {number}
         */
        get: function () {
            return this._progress;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebAudioTrack.prototype, "completed", {
        /**
         * Gets current track progress as a percentage
         *
         * @property completed
         * @readonly
         * @type {number}
         */
        get: function () {
            return this._completed;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebAudioTrack.prototype, "error", {
        /**
         * Gets any errors logged by HTML5 audio
         *
         * @property error
         * @readonly
         * @type {MediaError}
         */
        get: function () {
            return this.audio.error;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebAudioTrack.prototype, "canPlay", {
        /**
         * Gets a boolean value indicating whether the current source can be played
         *
         * @property canPlay
         * @readonly
         * @type {boolean}
         */
        get: function () {
            var format = "audio/" + this.audio.src.substr(this.audio.src.lastIndexOf('.') + 1);
            return this.audio && this.audio.canPlayType(format) != '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebAudioTrack.prototype, "isLoading", {
        /**
         * Gets a boolean value indicating whether the track is in loading state
         *
         * @property isLoading
         * @readonly
         * @type {boolean}
         */
        get: function () {
            return this._isLoading;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebAudioTrack.prototype, "hasLoaded", {
        /**
         * Gets a boolean value indicating whether the track has finished loading
         *
         * @property hadLoaded
         * @readonly
         * @type {boolean}
         */
        get: function () {
            return this._hasLoaded;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Plays current track
     *
     * @method play
     */
    WebAudioTrack.prototype.play = function () {
        if (!this.audio) {
            this.createAudio();
        }
        if (!this._hasLoaded) {
            console.log("Loading track " + this.src);
            this._isLoading = true;
        }
        //var source = this.ctx.createMediaElementSource(this.audio);  
        //source.connect(this.ctx.destination);
        this.audio.play();
    };
    /**
     * Pauses current track
     *
     * @method pause
     */
    WebAudioTrack.prototype.pause = function () {
        if (!this.isPlaying)
            return;
        console.log("Pausing track " + this.src);
        this.audio.pause();
        this.isPlaying = false;
    };
    /**
     * Stops current track and releases audio
     *
     * @method stop
     */
    WebAudioTrack.prototype.stop = function () {
        var _this = this;
        if (!this.audio)
            return;
        this.pause();
        this.audio.removeEventListener("timeupdate", function (e) { _this.onTimeUpdate(e); });
        this.isFinished = true;
        this.destroy();
    };
    /**
     * Seeks to a new position within the track
     *
     * @method seekTo
     * @param {number} time the new position to seek to
     */
    WebAudioTrack.prototype.seekTo = function (time) {
        this.audio.currentTime = time;
    };
    /**
     * Releases audio resources
     *
     * @method destroy
     */
    WebAudioTrack.prototype.destroy = function () {
        this.audio = undefined;
        console.log("Released track " + this.src);
    };
    return WebAudioTrack;
}());
WebAudioTrack = __decorate([
    Injectable(),
    __param(1, Optional()),
    __metadata("design:paramtypes", [String, String])
], WebAudioTrack);
export { WebAudioTrack };
